### DISCUSSION TITLE
Additional builtin languages support

### DISCUSSION THREAD

**Discussion by HerringtonDarkholme** [ast-grep/ast-grep#419] [General]
Currently ast-grep supports these languages: https://github.com/ast-grep/ast-grep/blob/main/crates/language/src/lib.rs
with notable languages listed [here](https://ast-grep.github.io/guide/introduction.html#supported-languages).

I believe there are more languages we need to support.

Please reply in comments below to vote for your favorite programming languages to be supported by ast-grep!
If your language is not listed, please create a new comment.

NOTE: Please do not create comments for duplicate language requests.

**Comment by HerringtonDarkholme** [comment]
CPP support:
* TIOBE index: 4
* Category: System Programming
* Tracking issue: https://github.com/ast-grep/ast-grep/issues/417
* tree-sitter Rust crate: https://crates.io/crates/tree-sitter-cpp
* crate status: seems to work

**Reply by HerringtonDarkholme** [reply]
CPP support is Done!

**Reply by HerringtonDarkholme** [reply]
Fixed!

**Comment by HerringtonDarkholme** [comment]
Ruby is supported!

**Comment by HerringtonDarkholme** [comment]
Julia Support!

**Reply by HerringtonDarkholme** [reply]
nope, julia's poularity unfortunately is not impactful enough for inclusion

**Comment by HerringtonDarkholme** [comment]
Solidity support? 


I hope N. Satoshi can wish my bitcoin call bullish. 🤡 

Unfortunately solidity does not have official tree-sitter support. (either from tree-sitter or solidity)

**Comment by john2x** [comment]
Terraform or HCL support would be amazing.

Is there documentation available for contributing a supported language?

**Reply by HerringtonDarkholme** [reply]
Thanks for suggestion.
Of course  there is contributing guide. https://ast-grep.github.io/contributing/add-lang.html


That said, I would think Terraform and HCL will be better supported by custom language registration? 

https://ast-grep.github.io/advanced/custom-language.html

**Reply by john2x** [reply]
Thanks. Wasn't aware of that capability. I'll take a look.

**Reply by thisguycodes** [reply]
@HerringtonDarkholme can you articulate why you feel a custom language registration is a better fit for HCL? I can understand wanting to keep scope down, but requiring folks to compile their own stuff is absolutely going to be a barrier to use (especially, but not exclusively, when people (me) want to use this in automated pipelines)

**Reply by HerringtonDarkholme** [reply]
@ThisGuyCodes ast-grep cannot bundle all languages in its binary. Built-in language should have enough community usage to deliver max impact with limited binary size.

If HCL has enough users upvote for inclusion in ast-grep, we will definitely consider it.

**Reply by HerringtonDarkholme** [reply]
Now it is reasonable to accept the HCL support PR, as many users are requesting it.

However, I will not work on it myself but expect community PR like 
https://github.com/ast-grep/ast-grep/pull/2107 and https://github.com/ast-grep/ast-grep/pull/2103

**Reply by HerringtonDarkholme** [reply]
HCL support is shipped in 0.39.3!

**Comment by kamal** [comment]
Glimmer (.hbs) would be great. There's a tree-sitter grammar at https://github.com/ember-tooling/tree-sitter-glimmer

**Comment by partner-pierre-vande-laar** [comment]
Some of our partners use [Ada](https://www.adacore.com/about-ada).
In the future, they might move to [Spark](https://www.adacore.com/about-spark).

**Reply by HerringtonDarkholme** [reply]
You can use https://ast-grep.github.io/advanced/custom-language.html

**Comment by pbsds** [comment]
Nix!

**Reply by HerringtonDarkholme** [reply]
Supported! https://github.com/ast-grep/ast-grep/pull/2107

**Comment by SteampunkIslande** [comment]
Snakemake ?
Its grammar is a superset of python, and there is already a linter available for it.
It would be great for automatic workflow migration

**Comment by VladimirMakaev** [comment]
How to add a custom language on all platforms? I.e what is the syntax to specify which parser to use on Linux, on Mac and Windows assuming I built 3 different shared libraries

**Reply by HerringtonDarkholme** [reply]
you can compile tree-sitter library as `parser.so` for all platforms

See ast-grep/langs script as example https://github.com/ast-grep/langs/blob/main/packages/angular/package.json#L7

Example artifacts https://github.com/ast-grep/langs/actions/runs/14602153038

**Reply by VladimirMakaev** [reply]
I completely understand that I can do that. My problem is that the ast-grep config lives in a monorepo that people are working with in multiple platforms. I.e. a user A gets a checkout on Windows and a user B gets a checkout on Linux, and user C gets a checkout on MacOS. Currently customLanguages/LibraryPath is just one file. So If I configure `parser.so` in sgconfig.yml compiled for Linux X64 then a user on Windows won't be able to configure `parser.dll` for Windows

**Reply by HerringtonDarkholme** [reply]
ast-grep at the moment does not support registering multiple binaries at the moment. instead, it relies on installation time binary picking up. https://github.com/ast-grep/langs

I wonder what is your setup. Are you commiting pre-compiled binary into the monorepo (fbcode?)  So your users will not have a installation time. 

I can give it a look if that's the use case.

**Reply by VladimirMakaev** [reply]
So for the ast-grep binary itself we do keep it precompiled in the repository. Every linter is effectively a binary that we have a contract with. People work with it with Macs, Windows, Linux machine and we expect exactly the same binary call to succeed on all platforms without users having to install anything on their machines. (Simplified it's `ast-grep scan --config <path_to_config>`). So everything is fine with built-in languages. If we register a custom language the config requires static configuration with concrete parser.so file which is not portable. At the moment we're trying to use ast-grep in the repository with LFS so we're ok to just add an .so file in the repository. But would like another solution if we were to allow custom languages in fbcode.

One way this can be fixed in my opinion is that if in addition to "libraryPath" we get "libraryLookupBinary" and ast-grep would execute it to get the path to the parser from stdout. Then the binary (most likely shell script) would be able to handle platform specifics and pick the right parser for the os/platform.

Also adding parameters to fully configure "customLanguages" via command line arguments can help too

**Reply by HerringtonDarkholme** [reply]
Thanks for the information. Using `libraryLookupBinary ` is probably too heavy, how about make `libraryPath` an object like 
```yaml
libraryPath:
  aarch64-apple-darwin: darwin.so
  aarch64-unknown-linux-gnu: linux-gnu.so
  i686-pc-windows-msvc: windows.so
```

So ast-grep can resolve the dynamic lib using the target triple defined by rust. https://github.com/dtolnay/target-triple

This will need some time for ast-grep to implement. At the same time, an alternative way is to maintain several different YAML configuration and having a wrapper script to invoke different binary like `ast-grep-aarch64-linux-gnu scan --config aarch64-linux-gnu.yml`.

**Reply by HerringtonDarkholme** [reply]
tracked by https://github.com/ast-grep/ast-grep/issues/2013

**Reply by HerringtonDarkholme** [reply]
shipped in 0.38.3

**Comment by seanf** [comment]
Please support Groovy! This would also be helpful for Jenkins CI (since `Jenkinsfile` is a form of Groovy)

(The frontrunner grammar seems to be https://github.com/murtaza64/tree-sitter-groovy, as used by nvim-treesitter.)

**Reply by HerringtonDarkholme** [reply]
generally ast-grep CLI now only accepts grammar from these sources:
* tree-sitter org  
* tree-sitter-grammar org
* official language org 
tree-sitter-groovy comes from an individual repo and ast-grep will wait for some official endorsement.
Plus, if Groovy is supported, should ast-grep support other build script language? say cmake. Or other config language like HCL

**Reply by seanf** [reply]
Groovy isn't just a build script language, it's another JVM language like Kotlin or Scala (but mostly dynamically typed). It just happens to have been used in Jenkins (and Gradle). But I take your point about official/trusted sources.

**Reply by seanf** [reply]
I have submitted https://issues.apache.org/jira/browse/GROOVY-11705 requesting an official tree-sitter grammar for Groovy.

**Comment by wont-work** [comment]
I'm not entirely sure if this would be relevant given it's not "code" per se, but supporting Markdown (or other markup formats?) could be interesting to let scripts (and AI? which I assume you'd be interested in given [it's advertised in the homepage](https://ast-grep.github.io/advanced/prompting.html)) filter files from headings and such.

Though, given how many dialects of Markdown there are, I'm not entirely sure how well this would work. I haven't really thought too deep about this.

**Reply by HerringtonDarkholme** [reply]
Unforntunately markdown's grammar is not too well-maintained. :/

https://github.com/tree-sitter-grammars/tree-sitter-markdown/issues

**Reply by HerringtonDarkholme** [reply]
Markdown is supported now!

**Comment by jayvdb** [comment]
I created a PR for adding TOML at https://github.com/ast-grep/ast-grep/pull/2661 .  

I note that https://ast-grep.github.io/contributing/add-lang.html does not mention that new languages must be suggested here in this discussion thread first.

**Comment by jayvdb** [comment]
Another suggestion is adding WIT support. There is a bit of a tree-sitter-wit problem that it would be useful to fix before doing this.  

The crate is owned by @Michael-F-Bryan , probably using code at https://github.com/Michael-F-Bryan/tree-sitter-wit , which is 35 commits behind the real source repo at https://github.com/bytecodealliance/tree-sitter-wit , which has its crate name also "tree-sitter-wit" c.f. https://github.com/bytecodealliance/tree-sitter-wit/blob/main/Cargo.toml , but has `repository` pointing to https://github.com/liamwh/tree-sitter-wit , which is a redirect back to https://github.com/bytecodealliance/tree-sitter-wit .

**Reply by Michael-F-Bryan** [reply]
This is a bit weird because back when I made my [WIT Language Server](https://github.com/Michael-F-Bryan/wit-lsp), there wasn't an official `tree-sitter-wit` repo. So I made my own and published it.

Later on, I found out about `liamwh/tree-sitter-wit` and [switched](https://github.com/Michael-F-Bryan/wit-lsp/pull/11) my language server over to their version instead (as a `git` dependency). As part of that, I ended up forking the repo and making some PRs to fix minor things.

In the 2 years since then, it looks like `liamwh/tree-sitter-wit` has been transferred to the bytecode alliance.

I'm more than happy to hand over full ownership of the `tree-sitter-wit` crate on crates.io to the bytecode alliance - I just need to know which team I need to give ownership to.

CC: https://github.com/bytecodealliance/tree-sitter-wit/issues/27#issuecomment-4524253397

**Comment by sleroq** [comment]
Svelte support would be useful.

As I understand it, .svelte files can be mapped to HTML as a workaround already, which handles some markup and script/style injections, but it does not understand Svelte template syntax like blocks, directives, special tags.

Native support would allow for rules like this for example:
```yaml
id: svelte-await-requires-catch
language: Svelte
message: Await blocks must handle rejected promises
severity: warning
rule:
  all:
    - kind: await_block
    - not:
        has:
          kind: branch_kind
          regex: "^catch$"
          stopBy: end
```

I have a draft implementation with [tree-sitter-htmlx-svelte](https://crates.io/crates/tree-sitter-htmlx-svelte) (supports Svelte 5 and is used by Helix editor). It adds about 37 KB compressed (+0.48%), keeping the binary around 7.73 MB. Would htmlx be acceptable despite being independently maintained?

There is also an alternative from `tree-sitter-grammars` and I'm happy to use it instead - [tree-sitter-svelte-ng](https://crates.io/crates/tree-sitter-svelte-ng) It's used in zed and neovim. However, based on the open issues it fails to parse `{@attach}` lacks newer `{let ...} / {const ...}` declaration tags (while HTMLX covers these features) and just seems unmaintained.

**Reply by sleroq** [reply]
On second thought, built-in support may not be necessary given Svelte's popularity and binary-size constraints. I'll just use custom language support instead.

Thanks for pointing me to the documentation, I should have read it more carefully before opening the PR.

**Reply by HerringtonDarkholme** [reply]
Thanks for your understanding 🙏
